File: Diagrammatize.Design.Pattern/code/cxx/Decorator/A2/MultiStringDisplay.cpp
#include "MultiStringDisplay.hpp"

void MultiStringDisplay::add(const std::string& msg) {
    body.push_back(msg);
    updateColumn(msg);
}

int MultiStringDisplay::getColumns() {
    return columns;
}

int MultiStringDisplay::getRows() {
    return body.size();
}

std::string MultiStringDisplay::getRowText(int row) {
    return body[row];
}

void MultiStringDisplay::updateColumn(const std::string& msg) {
    if (msg.length() > columns) {
        columns = msg.length();
    }
    for (int row = 0; row < body.size(); row++) {
        int fills = columns - body[row].length();
        if (fills > 0) {
            body[row] = body[row] + spaces(fills);
        }
    }
}

std::string MultiStringDisplay::spaces(int count) {
    return std::string(count, ' ');
}

File: Diagrammatize.Design.Pattern/code/python/Decorator/A2/MultiStringDisplay.py
from Display import Display

class MultiStringDisplay(Display):
    def __init__(self):
        self.body = list()
        self.columns = 0

    def add(self, msg):
        self.body.append(msg)
        self.updateColumn(msg)

    def getColumns(self):
        return self.columns

    def getRows(self):
        return len(self.body)

    def getRowText(self, row):
        return self.body[row]

    def updateColumn(self, msg):
        if len(msg) > self.columns:
            self.columns = len(msg)

        for row in range(len(self.body)):
            fills = self.columns - len(self.body[row])
            if fills > 0:
                self.body[row] = self.body[row] + self.spaces(fills)

    def spaces(self, count):
        return ' '*count
